hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java

```diff
@@ -398,49 +370,132 @@ private void updateExpiryFields(
+        if (updatesAdminOnlyNonKeyTokenProperty(op)) {
+            requireAdmin(context, token);
         }
-        if (op.hasMetadataKey()) {
-            validateTrue(originalToken.hasMetadataKey(), TOKEN_HAS_NO_METADATA_KEY);
-            builder.metadataKey(op.metadataKey());
+        // Any role key can be updated if the key is not empty of the token.
+        // Updating any non-admin keys with the key verification mode is NO_VALIDATION needs either
+        // admin key or the role key to sign.If the key verification mode is FULL_VALIDATION, then
+        // the new key also should sign the transaction.
+        for (final var tokenKey : NON_ADMIN_TOKEN_KEYS) {
+            if (tokenKey.isPresentInUpdate(op)) {
+                final var newRoleKey = tokenKey.getFromUpdate(op);
+                if (!isKeyRemoval(newRoleKey)) {
+                    if (op.keyVerificationMode() == NO_VALIDATION) {
+                        requireAdminOrRole(context, token, tokenKey);
+                    } else {
+                        // With "full" verification mode, our required key
+                        // structure is a 1/2 threshold with components:
+                        //   - Admin key
+                        //   - A 2/2 list including the role key and its replacement key
+                        final var key = tokenKey.getFromUpdate(op);
+                        requireAdminOrRole(context, token, tokenKey, key);
+                    }
+                }
+            }
         }
+    }
 
-        if (isMetadataOnlyUpdateOp(op)) {
-            validateTrue(originalToken.hasAdminKey() || originalToken.hasMetadataKey(), TOKEN_IS_IMMUTABLE);
-        } else if (!isExpiryOnlyUpdateOp(op)) {
-            validateTrue(originalToken.hasAdminKey(), TOKEN_IS_IMMUTABLE);
+    private void requireAdminOrRole(
+            @NonNull final PreHandleContext context, @NonNull final Token token, @NonNull final TokenKey roleKey)
+            throws PreCheckException {
+        requireAdminOrRole(context, token, roleKey, null);
+    }
+
+    private void requireAdminOrRole(
+            @NonNull final PreHandleContext context,
+            @NonNull final Token token,
+            @NonNull final TokenKey roleKey,
+            @Nullable final Key replacementKey)
+            throws PreCheckException {
+        final var maybeRoleKey = roleKey.getFromToken(token);
+        // Prioritize TOKEN_IS_IMMUTABLE for completely immutable tokens
+        mustExist(maybeRoleKey, token.hasAdminKey() ? roleKey.tokenHasNoKeyStatus() : TOKEN_IS_IMMUTABLE);
+        if (token.hasAdminKey()) {
+            context.requireKey(oneOf(
+                    replacementKey == null ? maybeRoleKey : allOf(maybeRoleKey, replacementKey),
+                    token.adminKeyOrThrow()));
+        } else {
+            context.requireKey(maybeRoleKey);
+            if (replacementKey != null) {
+                context.requireKey(replacementKey);
+            }
```

Comment:
If the replacementKey is required to sign and its not valid, it will not be added to the requiredNonPayerKeys. We should validate and throw here ?

Reply:
Ah, yes we validate that in `pureChecks`